funsor/constant.py

```diff
@@ -0,0 +1,158 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import OrderedDict
+from functools import reduce
+
+import funsor.ops as ops
+from funsor.tensor import Tensor
+from funsor.terms import (
+    Binary,
+    Funsor,
+    FunsorMeta,
+    Number,
+    Reduce,
+    Unary,
+    Variable,
+    eager,
+)
+
+
+class ConstantMeta(FunsorMeta):
+    """
+    Wrapper to convert ``const_inputs`` to a tuple.
+    """
+
+    def __call__(cls, const_inputs, arg):
+        if isinstance(const_inputs, dict):
+            const_inputs = tuple(const_inputs.items())
+        return super(ConstantMeta, cls).__call__(const_inputs, arg)
+
+
+class Constant(Funsor, metaclass=ConstantMeta):
+    """
+    Funsor that is constant wrt ``const_inputs``.
+
+    :class:`Constant` can be used for provenance tracking.
+
+    Examples::
+
+        a = Constant(OrderedDict(x=Real, y=Bint[3]), Number(0))
+        a(y=1)  # returns Constant(OrderedDict(x=Real), Number(0))
+        a(x=2, y=1)  # returns Number(0)
+
+        d = Tensor(torch.tensor([1, 2, 3]))["y"]
+        a + d  # returns Constant(OrderedDict(x=Real), d)
+
+        c = Constant(OrderedDict(x=Bint[3]), Number(1))
+        c.reduce(ops.add, "x")  # returns Number(3)
+
+    :param dict const_inputs: A mapping from input name (str) to datatype (``funsor.domain.Domain``).
+    :param funsor arg: A funsor that is constant wrt to const_inputs.
+    """
+
+    def __init__(self, const_inputs, arg):
+        assert isinstance(arg, Funsor)
+        assert isinstance(const_inputs, tuple)
+        const_inputs = OrderedDict(const_inputs)
+        # const_inputs and arg.inputs have to be disjoint
+        assert set(const_inputs).isdisjoint(arg.inputs)
+        inputs = const_inputs.copy()
+        inputs.update(arg.inputs)
+        output = arg.output
+        fresh = frozenset(const_inputs)
+        bound = {}
+        super(Constant, self).__init__(inputs, output, fresh, bound)
+        self.arg = arg
+        self.const_vars = frozenset(Variable(k, v) for k, v in const_inputs.items())
+        self.const_inputs = const_inputs
+
+    def eager_subs(self, subs):
+        assert isinstance(subs, tuple)
+        subs = OrderedDict((k, v) for k, v in subs)
+        const_inputs = OrderedDict()
+        for k, d in self.const_inputs.items():
+            if k in subs:
+                v = subs[k]
+                const_inputs.update(
+                    (name, value)
+                    for name, value in v.inputs.items()
```

Comment:
Check that `v.inputs` doesn't overlap with `self.arg.inputs`?

Reply:
This substitution behavior is similar to how you would index expanded `Tensor`. For example:

```py
a = Tensor(torch.tensor([1, 2]))["y"]
b = Tensor(torch.ones(3, 2))["x", "y"]  # expanded Tensor
b(x=a)  # allowed
c = Constant(OrderedDict(x=Bint[3]), Tensor(torch.ones(2))["y"])
c(x=a)  # y is overlapping; similar to b(x=a)
```
I would vote to allow overlapping inputs but add an assertion:

```py
assert all(v.inputs[k] == self.arg.inputs[k] for k in set(v.inputs).intersection(self.arg.inputs))
```